lang/src/splgen/splgen/contract_generator.py

```diff
@@ -50,33 +72,26 @@ def generate(model: Script, output_file: str):
 
     for ca in manager.contract_administrators:
         for method in ca.methods:
-            contracts[ca.contract.name].functions[method.name].mod.add(
+            contracts[ca.contract.name].functions[method.name].modifiers.add(
                 manager.method.name)
 
     for admin in model.administration.extension_administrators:
         for ca in admin.contract_administrators:
             for method in ca.methods:
-                contracts[ca.contract.name].functions[method.name].mod.add(
+                contracts[ca.contract.name].functions[method.name].modifiers.add(
                     admin.method.name)
 
     for contract in contracts.values():
-        for fun in contract.functions.values():
-            fun.insert_modifiers()
-
-    with open(output_file, 'w') as f:
-        f.write(f"/* Generated by SPL [{datetime.now()}] */\n\n")
-        f.write("// SPDX-License-Identifier: MIT\npragma solidity ^0.8.17;\n\n")
-        f.write("//"+" DEPENDENCIES ".center(100, '=')+'\n\n')
-        for d in sorted(list(dependencies), key=lambda d: (d.type, d.priority), reverse=True):
-            f.write(d.code+'\n')
-        f.write("//"+" YOUR CONTRACTS ".center(100, '=')+'\n\n')
-        for c in contracts.values():
-            f.write(f"contract {c.name} " + "{\n")
-            for d in sorted(list(c.dependencies), key=lambda d: d.type, reverse=True):
-                f.write('\t'+d.code+'\n')
-            for m in c.functions.values():
-                f.write('\t'+m.code+'\n')
-            f.write("\n}\n")
+        contract.remove_redundant_dependencies()
+        contract.insert_modifiers()
+
+    env = Environment(loader=BaseLoader)
+    env.filters["sort_deps"] = lambda deps: sorted(list(deps), key=lambda d: (d.type, d.priority), reverse=True)
+    env.from_string(template).stream(
+        dependencies=dependencies,
+        contracts=contracts.values(),
+        timestamp=datetime.now()
```

Comment:
How is this serialized? Is it a nice human-readable format?

Reply:
Like this: 2023-03-20 20:07:05.271160

Reply:
Not the nicest format but we can go with that one 😄